modules/apps/29-fee/ibc_middleware.go

```diff
@@ -401,3 +396,30 @@ func (IBCMiddleware) UnwrapVersionUnsafe(version string) (string, string, error)
 
 	return metadata.FeeVersion, metadata.AppVersion, nil
 }
+
+// UnwrapVersionSafe unwraps a version contextually by relying on storage and the given portID and channelID.
+func (im IBCMiddleware) UnwrapVersionSafe(ctx sdk.Context, portID, channelID, version string) (string, string) {
+	if !im.keeper.IsFeeEnabled(ctx, portID, channelID) {
+		return "", version
+	}
+	metadata, err := types.MetadataFromVersion(version)
+	if err != nil {
+		// This should not happen, as it would mean that the channel is broken. Only a severe bug would cause this.
+		panic(errorsmod.Wrap(err, "failed to unwrap app version from channel version"))
+	}
+	return metadata.FeeVersion, metadata.AppVersion
+}
+
+// UnwrapAcknowledgement unwraps an acnkowledgement contextually by relying on storage and the given portID and channelID.
+func (im IBCMiddleware) UnwrapAcknowledgement(ctx sdk.Context, portID, channelID string, ack []byte) ([]byte, []byte) {
+	if !im.keeper.IsFeeEnabled(ctx, portID, channelID) {
+		return nil, ack
+	}
+
+	var incentivizedAck types.IncentivizedAcknowledgement
+	if err := json.Unmarshal(ack, &incentivizedAck); err != nil {
+		panic(errorsmod.Wrap(err, "failed to unwrap acknowledgement"))
+	}
+
+	return ack, incentivizedAck.AppAcknowledgement
```

Comment:
note: we return ack we unmarshaled as our ack. This is because it expects to unmarshal this value in `OnAcknowledgement`